core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -115,10 +119,18 @@
     private final AtomicInteger throttleRequestCount = new AtomicInteger();
     private final EngineConfig.OpenMode openMode;
     private final AtomicBoolean allowCommits = new AtomicBoolean(true);
+    private final AtomicLong highestDeOptimizeAddDocumentTimestamp = new AtomicLong(-1); // TODO maybe we can find a shorter name
+    private final CounterMetric numVersionLookups = new CounterMetric();
+    private final CounterMetric numVersionsLoaded = new CounterMetric();
 
     public InternalEngine(EngineConfig engineConfig) throws EngineException {
         super(engineConfig);
         openMode = engineConfig.getOpenMode();
+        if (engineConfig.getOptimizeAutoGeneratedIds() == false
+            || engineConfig.getIndexSettings().getIndexVersionCreated().before(Version.V_5_0_0_alpha6)) {
+            // no optimization for pre 5.0.0.alpha6 since translog might not have all information needed
```

Comment:
Sorry, I forget to respond to this earlier, but ops coming from the old translog won't have the time stamp and therefore will not be optimized. They can also not compete with new concurrent incoming request because they will have a different id (and we flush at the end of primary recovery, but we shouldn't rely on this).